compiler/rustc_hir_analysis/src/astconv/mod.rs

```diff
@@ -2405,13 +2405,119 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
+        let mut error_occured = false;
+        let mut try_emit = |error| {
+            if emit {
+                self.tcx().sess.emit_err(error);
+            }
+            error_occured = true;
+        };
+        let res_span = self.tcx().def_span(res_id);
+
+        let res_generics = self.tcx().generics_of(res_id);
+        let parent = self.tcx().parent(self.item_def_id());
+        let parent_generics = self.tcx().generics_of(parent);
+        let self_ty_bias = (self.tcx().def_kind(parent) == DefKind::Trait) as usize;
+        if parent_generics.count() + res_generics.own_counts().count() > self_ty_bias {
+            try_emit(super::errors::NotSupportedDelegation {
+                span,
+                descr: "early bound generics",
+                res_span,
+            });
+        }
+        if self.tcx().asyncness(res_id) == ty::Asyncness::Yes {
+            try_emit(super::errors::NotSupportedDelegation {
+                span,
+                descr: "async functions",
+                res_span,
+            });
+        }
+
+        if self.tcx().constness(res_id) == hir::Constness::Const {
+            try_emit(super::errors::NotSupportedDelegation {
+                span,
+                descr: "const functions",
+                res_span,
+            });
+        }
+
+        let res_sig = self.tcx().fn_sig(res_id).instantiate_identity();
+        if res_sig.c_variadic() {
+            try_emit(super::errors::NotSupportedDelegation {
+                span,
+                descr: "variadic functions",
+                res_span,
+            });
+            return error_occured;
+        }
+
+        if let hir::Unsafety::Unsafe = res_sig.unsafety() {
+            try_emit(super::errors::NotSupportedDelegation {
+                span,
+                descr: "unsafe functions",
+                res_span,
+            });
+        }
+
+        if abi::Abi::Rust != res_sig.abi() {
+            try_emit(super::errors::NotSupportedDelegation {
+                span,
+                descr: "non Rust ABI functions",
+                res_span,
+            });
+        }
+
+        error_occured
+    }
+
+    fn ty_from_delegation(
+        &self,
+        res_id: DefId,
+        idx: hir::InferDelegationVar,
+        span: Span,
+    ) -> Ty<'tcx> {
+        if self.check_delegation_constraints(res_id, span, idx == hir::InferDelegationVar::Output) {
+            let e = self.tcx().sess.span_delayed_bug(span, "not supported delegation case");
+            self.set_tainted_by_errors(e);
+            return Ty::new_error(self.tcx(), e);
+        };
+        let res_sig = self.tcx().fn_sig(res_id);
+
+        let parent = self.tcx().parent(self.item_def_id());
+        let item_def_kind = self.tcx().def_kind(self.item_def_id());
+        let parent_def_kind = self.tcx().def_kind(parent);
+
+        let res_sig = match (item_def_kind, parent_def_kind) {
```

Comment:
```suggestion
        let res_sig = match parent_def_kind {
```
`item_def_kind` doesn't affect anything here.